### ISSUE TITLE
[Core] Restore worker silently fails and the program is stuck

### ISSUE THREAD

**Issue by frank-lsf** [ray-project/ray#24248]
### What happened + What you expected to happen

When the restore worker encounters an error (e.g. in #24196, or in general for example the spilled file is removed), it silently fails (only prints an error `/tmp/ray/session_latest/logs/io-*.err`). It looks like the system will schedule a few retries to restore the object, but eventually give up.

The behavior on the application side is that the program is stuck, and no progress is made, and no error is thrown, nor is there any error message. It would be nice if the error can be popped back to the application. For example, when the spill worker errors due to disk space or other reason, the application receives a Python error. Can we do the same for restore workers?

### Versions / Dependencies

master

### Reproduction script

You can manually make the restore worker throw in `external_storage.py`, then test its behavior.

### Issue Severity

Low: It annoys or frustrates me.

**Comment by stephanie-wang** [comment]
@rkooo567 do you think you could take this one since you worked on the spill error?

**Comment by rkooo567** [comment]
Yeah I'd love to. It seems to be a part of better error message as well (which I have worked a bit and stopped for a while). Is P2 the correct priority?

**Comment by stephanie-wang** [comment]
I'm upgrading this to P1 since it is actually pretty bad for debugging if we're silently dropping errors.

@rkooo567 do you think you have bandwidth to take this in the next week or two?

**Comment by rkooo567** [comment]
Hmm I will be a bit busy until June 21st I think. I can share some thoughts if there's anybody who has bandwidth on Shuffle project's side instead?

**Comment by stephanie-wang** [comment]
I think we can find someone to work on it. But please share what you know about the issue so far!

**Comment by kennethlien** [comment]
> Yeah I'd love to. It seems to be a part of better error message as well (which I have worked a bit and stopped for a while). 

Could you point us to the related PRs / any examples of bubbling up errors from workers?

**Comment by rkooo567** [comment]
@kennethlien 
There are two paths where things can hang forever.

Spilling: needed when ray.put is called
Restoration: needed when ray.get is called

Right now, if spilling is failed, it is retried later when another spilling happens. (Ideally, we'd like to fail ray.put if this happens a lot, but I think this is P2). 

For restoration, we have the hanging issue described in this issue. 

Basically, this is how the problem happens;
- When the object is not available from the object store, and we need to restore, pull_manager pulls them. It uses the restore worker to restore the object (Look `external_stroage.py` to see how the restoration works). https://github.com/ray-project/ray/blob/c74886a55e81dc8f4150094bfec69d875f0b365d/src/ray/object_manager/pull_manager.cc#L531
- If restoration failed, the restore worker just return ok status. Since this won't actually restore the object, pull manager retries to restore the object (by default, it retries pulling 30 seconds + exponential backoff if the object is not available in the next retry timer). https://github.com/ray-project/ray/blob/c74886a55e81dc8f4150094bfec69d875f0b365d/python/ray/_raylet.pyx#L935
- It means that if it keeps failing to restore objects, it will be retried forever with 30 seconds * exponential backoff delay (which will eventually become like hanging).

Ideally, we'd like to catch failures and raise an exception for users when restoration keeps failing.

what you need to do is; 

1. When the restoration request is failed from restore worker, we'd like to return failed status + exception message to the raylet instead of okay status. https://github.com/ray-project/ray/blob/c74886a55e81dc8f4150094bfec69d875f0b365d/python/ray/_raylet.pyx#L935
2. If the restoration request is failed, https://github.com/ray-project/ray/blob/c74886a55e81dc8f4150094bfec69d875f0b365d/src/ray/object_manager/pull_manager.cc#L535, instead of doing nothing, we should record # of retries (it could be 0 if we'd like to simplify the implementation). If retries == max retries, we just need to mark the object as failed using `MarkObjectsAsFailed` from `node_manager.cc` (you should pass this method as a callback to pull manager). Example: https://github.com/ray-project/ray/blob/c74886a55e81dc8f4150094bfec69d875f0b365d/src/ray/raylet/node_manager.cc#L228. This creates a Ray object on the corresponding object id, so that the caller can raise an exception. https://github.com/ray-project/ray/blob/c74886a55e81dc8f4150094bfec69d875f0b365d/python/ray/serialization.py#L263
3. Note that you can provide additional data to the error object (the metadata == error type, and the data can contain actual exception messages so that users can see why things are failing). This is how we create a ray object from the core worker to provide actual data (by creating a ray object with metadata as specified here https://github.com/ray-project/ray/blob/c74886a55e81dc8f4150094bfec69d875f0b365d/src/ray/core_worker/task_manager.cc#L601). Note that providing data to the error object is not supported by Raylet yet. You can talk to me to figure out how to support it, but it is also okay to not include the data in the short term (I will work on improving this path anyway). If you can provide the data, you can include the exception message here. 
4. Now ray.get will raise an exception using this error object.

**Comment by kennethlien** [comment]
Working on this at https://github.com/ray-project/ray/pull/25000

**Comment by kennethlien** [comment]
> When the restoration request is failed from restore worker, we'd like to return failed status + exception message to the raylet instead of okay status.

Is there a way to do this without changing the return value of cdef int64_t restore_spilled_objects_handler(



> 1. ray/python/ray/_raylet.pyx

**Comment by franklsf95** [comment]
@kennethlien I think here you can just throw the exception like we do in the `spill_objects_handler` function above.

**Comment by kennethlien** [comment]
@rkooo567 We're not sure how to get `restore_spilled_objects_handler` to return the error without large changes. Would it be reasonable to mark the object as failed after a certain number of retries?

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
